server/src/main/java/org/apache/iotdb/db/mpp/buffer/SourceHandle.java

```diff
@@ -19,72 +19,389 @@
 
-  private long getRetainedSizeInBytes(TsBlock tsBlock) {
-    throw new UnsupportedOperationException();
-  }
+  private synchronized void trySubmitGetDataBlocksTask() {
+    final int startSequenceId = nextSequenceId;
+    int endSequenceId = nextSequenceId;
+    long reservedBytes = 0L;
+    ListenableFuture<?> future = null;
+    while (sequenceIdToDataBlockSize.containsKey(endSequenceId)) {
+      Long bytesToReserve = sequenceIdToDataBlockSize.get(endSequenceId);
+      if (bytesToReserve == null) {
+        throw new IllegalStateException("Data block size is null.");
+      }
+      future =
+          localMemoryManager
+              .getQueryPool()
+              .reserve(localFragmentInstanceId.getQueryId(), bytesToReserve);
+      if (future.isDone()) {
+        endSequenceId += 1;
+        reservedBytes += bytesToReserve;
+        bufferRetainedSizeInBytes += bytesToReserve;
+      } else {
+        break;
+      }
+    }
 
-  @Override
-  public boolean isFinished() {
-    return finished;
+    if (future == null) {
+      // Next data block not generated yet. Do nothing.
+      return;
+    }
+
+    if (future.isDone()) {
+      nextSequenceId = endSequenceId;
+      executorService.submit(new GetDataBlocksTask(startSequenceId, endSequenceId, reservedBytes));
+      numActiveGetDataBlocksTask += 1;
+    } else {
+      nextSequenceId = endSequenceId + 1;
+      // The future being not completed indicates,
+      //   1. Memory has been reserved for blocks in [startSequenceId, endSequenceId).
+      //   2. Memory reservation for block whose sequence ID equals endSequenceId is blocked.
+      //   3. Have not reserve memory for the rest of blocks.
+      //
+      //  startSequenceId             endSequenceId  endSequenceId + 1
+      //         |-------- reserved --------|--- blocked ---|--- not reserved ---|
+
+      if (endSequenceId > startSequenceId) {
+        // Memory has been reserved. Submit a GetDataBlocksTask for these blocks.
+        executorService.submit(
+            new GetDataBlocksTask(startSequenceId, endSequenceId, reservedBytes));
+        numActiveGetDataBlocksTask += 1;
+      }
+
+      // Submit a GetDataBlocksTask when memory is freed.
+      final int sequenceIdOfUnReservedDataBlock = endSequenceId;
+      final long sizeOfUnReservedDataBlock = sequenceIdToDataBlockSize.get(endSequenceId);
+      future.addListener(
+          () -> {
+            executorService.submit(
+                new GetDataBlocksTask(
+                    sequenceIdOfUnReservedDataBlock,
+                    sequenceIdOfUnReservedDataBlock + 1,
+                    sizeOfUnReservedDataBlock));
+            numActiveGetDataBlocksTask += 1;
+            bufferRetainedSizeInBytes += sizeOfUnReservedDataBlock;
+          },
+          executorService);
+
+      // Schedule another call of trySubmitGetDataBlocksTask for the rest of blocks.
+      future.addListener(SourceHandle.this::trySubmitGetDataBlocksTask, executorService);
```

Comment:
If we schedule another `trySubmitGetDataBlocksTask` task once the future is done. Is `line 181-193` not necessary ?

And on the other hand, if `trySubmitGetDataBlocksTask` runs before the 186, the total memory reserved may exceed the threshold.

Reply:
MemoryPool will increase the reserved amount of memory before marking the future as completed. Thus, the next call of `trySubmitGetDataBlocksTask ` is safe.